File: monty.py
#!/usr/bin/python3

import random
import sys
import time

iteracoes = int(sys.argv[1])
estrategia = str(sys.argv[2])

vitoria = 0
perdicao = 0

start_time = time.time()

for i in range(iteracoes):
    premios = ["cabra", "cabra", "cabra"]
    print("----------JOGO", i+1, "----------")

    portaCarro = random.randint(0,2)
    portaJogador = random.randint(0,2)

    premios[portaCarro] = "carro"

    print("Voce escolheu a porta", portaJogador+1)

    for j in range(0,3):
        if (j != portaCarro and j != portaJogador):
            print("A porta", j+1, "tem uma cabra")
            portaCabra = j
            break

    if estrategia == 't':
        for k in range(0,3):
            if (portaJogador != k and portaCabra != k):
                portaJogador = k
                break

        print("Voce troca para a porta", portaJogador+1)
    else:
        print("Voce continua na porta", portaJogador+1)


    if (portaJogador == portaCarro):
        print("A sua porta tem um carro! PARABENS!")
        vitoria = vitoria + 1
    else:
        print("A sua porta tem uma cabra! Mais sorte na proxima")
        perdicao = perdicao + 1

    print(" ")

    print(premios[0], "|", premios[1], "|", premios[2])

    print(" ")


end_time = time.time()

print("---------\n")
print("Tempo de exec:", end_time - start_time, "segundos")
print(vitoria, "vitorias -----> ", (vitoria * 100) / iteracoes, "%")
print(perdicao, "derrotas -----> ", (perdicao * 100) / iteracoes, "%")

File: monty.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#define NUMERO_PORTAS 3

void printaPorta(int tipoPorta); void trocaCor(int n);

int main(int argc, char *argv[]) {

  //sem argumentos o programa nao continua
  argc > 2 ? : exit(0);

  long int iteracoes;
  long int vitoria = 0;
  long int perdicao = 0;
  int flagPremios[NUMERO_PORTAS];
  char estrategia;
  int portaJogador, portaCarro;
  int portasAbertas;

  srand(time(NULL));
  int seed = rand() % 100;

  iteracoes = atoi(argv[1]);
  estrategia = argv[2][0];

  clock_t begin = clock();

  for (long int i = 0; i < iteracoes; i++) {
    portasAbertas = 0;
    for (int j = 0; j < NUMERO_PORTAS; j++) {
      flagPremios[j] = 0;
    }

    printf("----------JOGO %ld----------\n", i+1);

    srandom(seed++);
    portaCarro = random() % NUMERO_PORTAS;
    srandom(seed * 3 + 1);
    portaJogador = random() % NUMERO_PORTAS;

    flagPremios[portaCarro] = 2;

    printf("Voce escolheu a porta %d\n", portaJogador+1);

    while (portasAbertas != NUMERO_PORTAS - 2) {

      for (int j = 0; j < NUMERO_PORTAS; j++) {
        if (j != portaCarro && j != portaJogador && flagPremios[j] != 1) {
          printf("A porta %d tem uma cabra\n", j+1);
          portasAbertas++;
          flagPremios[j] = 1;
          break;
        }
      }

      if (estrategia == 't') {
        for (int j = 0; j < NUMERO_PORTAS; j++) {
          if (portaJogador != j && flagPremios[j] != 1) {
            portaJogador = j;
            break;
          }
        }
        printf("Voce troca para a porta %d\n", portaJogador+1);
      } else {
        printf("Voce fica com a porta %d\n", portaJogador+1);
      }
    }


    if (flagPremios[portaJogador] == 2) {
      printf("\033[1;32m"); //verde
      printf("A sua porta tem um carro! PARABENS!\n");
      printf("\033[0m");    //reseta
      vitoria++;
    } else {
      printf("\033[1;31m"); //vermelho
      printf("A sua porta tem uma cabra! Mais sorte na proxima\n");
      printf("\033[0m");    //reseta
      perdicao++;
    }


    puts("");

    for (int j = 0; j < NUMERO_PORTAS; j++) {
      trocaCor(2);
      printaPorta(flagPremios[j]);
      trocaCor(0);
      printf(" | ");
    }

    puts("\n");

  }

  clock_t end = clock();

  double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;

  puts("----------");

  printf("Tempo da iteracao: %.3f segundos\n", time_spent);
  trocaCor(1);
  printf("%ld vitorias -----> %ld%% \n", vitoria, (vitoria * 100) / iteracoes);
  trocaCor(2);
  printf("%ld derrotas -----> %ld%% \n", perdicao, (perdicao * 100) / iteracoes);
  trocaCor(0);
  printf("\n");

  return 0;
}

void printaPorta(int tipoPorta){
  switch (tipoPorta) {
    case 0:
      printf("cabra");
      break;

    case 1:
      printf("cabra");
      break;

    case 2:
      trocaCor(1);
      printf("carro");
      break;
  }
}

void trocaCor(int n) {
  switch (n) {
    case 0:
      printf("\033[0m");    //reseta
      break;
    case 1:
      printf("\033[1;32m"); //verde
      break;
    case 2:
      printf("\033[1;31m"); //vermelho
      break;
  }
}
